pkg/api/jobrunintervals/gcsscanner.go

```diff
@@ -0,0 +1,216 @@
+	}
+	resp, err := netClient.Get(url)
+	if err != nil {
+		return nil, fmt.Errorf("failed to fetch %s: %v", url, err)
+	}
+	defer resp.Body.Close()
+
+	z := html.NewTokenizer(resp.Body)
+	for {
+		tt := z.Next()
+
+		switch {
+		case tt == html.ErrorToken:
+			// End of the document, we're done
+			return links, nil
+		case tt == html.StartTagToken:
+			t := z.Token()
+
+			isAnchor := t.Data == "a"
+			if isAnchor {
+				for _, a := range t.Attr {
+					if a.Key == "href" {
+						links = append(links, a.Val)
+						break
+					}
+				}
+			}
+		}
+	}
+}
+
+func GetMatchingLinkFromURL(baseURL string, regex *regexp.Regexp, onlyMatchLastSegment bool) (*url.URL, error) {
+	urls, err := GetMatchingLinksFromURL(baseURL, []*regexp.Regexp{regex}, onlyMatchLastSegment)
+	if err != nil {
+		return nil, err
+	}
+	// We're expecting exactly one in this function call, anything else is an error:
+	if len(urls) != 1 {
+		return nil, fmt.Errorf("expected 1 matching URL, found %d on: %s", len(urls), baseURL)
+	}
+	return urls[0], nil
+}
+
+// GetMatchingLinksFromURL scans all the links found on the given URL for those matching a regex.
+// Used to crawl GCS bucket links from our prow jobs.
+func GetMatchingLinksFromURL(baseURL string, regexes []*regexp.Regexp, onlyMatchLastSegment bool) ([]*url.URL, error) {
+	allLinks, err := GetLinksFromURL(baseURL)
+	if err != nil {
+		return []*url.URL{}, fmt.Errorf("failed to fetch links on %s: %v", baseURL, err)
+	}
+	if len(allLinks) == 0 {
+		return []*url.URL{}, fmt.Errorf("no links found on: %s", baseURL)
+	}
+
+	matchedLinks := []string{}
+	for _, link := range allLinks {
+		log.WithField("link", link).Debug("checking link")
+		linkSplitBySlash := strings.Split(link, "/")
+		lastPathSegment := linkSplitBySlash[len(linkSplitBySlash)-1]
+		if len(lastPathSegment) == 0 {
+			lastPathSegment = linkSplitBySlash[len(linkSplitBySlash)-2]
+		}
+		for _, re := range regexes {
+			if (onlyMatchLastSegment && re.Match([]byte(lastPathSegment))) ||
+				(!onlyMatchLastSegment && re.Match([]byte(link))) {
+
+				log.WithField("link", link).Debug("found link match")
+				matchedLinks = append(matchedLinks, link)
+			}
+		}
+	}
+
+	matchedURLs := make([]*url.URL, len(matchedLinks))
+	for i, ml := range matchedLinks {
+		matchedLink := ml
+		if strings.HasPrefix(ml, "/") {
+			matchedLink = gcsPrefix + ml
+		}
+		mURL, err := url.Parse(matchedLink)
+		if err != nil {
+			return []*url.URL{}, fmt.Errorf("failed to parse URL from link %s: %v", ml, err)
+		}
+		matchedURLs[i] = mURL
+
+	}
+	return matchedURLs, nil
+}
+
+func ReadIntervalsFile(url *url.URL) (*api.EventIntervalList, error) {
+	log.WithField("url", url).Info("getting intervals file")
+
+	// Get the base filename we'll add to each incoming interval:
+	tokens := strings.Split(url.Path, "/")
+	baseFile := tokens[len(tokens)-1]
+
+	resp, err := http.Get(url.String())
+	if err != nil {
+		return nil, err
+	}
+
+	defer resp.Body.Close()
+
+	body, err := ioutil.ReadAll(resp.Body)
```

Comment:
How big are the intervals files? You could use Go's built in ReverseProxy to just stream the data https://pkg.go.dev/net/http/httputil#ReverseProxy

Reply:
They're around 15-30M today, I've switched the code over to the existing GCS stuff though so hopefully this isn't a concern, or one we can address if it becomes an issue.